### PR TITLE
Fixing the mechanics of sources

### LINKED ISSUE #595: Sources - Wrong times for frequency
I tested an other function for the sources. I have 5 sources which have the same frequency, but different start times (time_min). So I would expect, that the sources start generating agents at time_min and the next time is time_min+frequency. But what happens its, that each source starts at time_min and then generates the agents with a "global" frequency, so there is no difference between the times when agents are created from different sources. You can see it on the picture. "Soll" = what is expected, "ist" = what happens in the simulation.

Bild
Here are the files: [Archiv.zip](https://github.com/JuPedSim/jpscore/files/3950980/Archiv.zip)

- [x] Fix Bug
- [x] Create Tests to detect this bug in the future

### PR DESCRIPTION
**Pull request by chraibi**
`rate` of a source defaults to 1. 
This is wrong and leads creating agents at the wrong time, then 
[subCycle](https://github.com/JuPedSim/jpscore/blob/master/libcore/src/pedestrian/AgentsSourcesManager.cpp#L107) may be true when it should not be.

Not defining `rate` means actually we create all `N_create` at once. 

Therefore, in this PR `rate` defaults to `frequency`. 

- [x] better source defaults
- [x] update documentation (waiting for #626)
- [x] agent's creation should start at `time_min`
- [x] add a test

Fixes #595

### CODE REVIEW

**Comment by zeroset** [comment]
Why do you call the tests develop_test? What do you mean with develop?

**Comment by zeroset** [comment]
Please add the required python dependencies to `container/build/Dockerfile`

**Comment by chraibi** [comment]
> Why do you call the tests develop_test? What do you mean with develop?

These are featurs still WIP.
(being **develop**ed)

But probably we can move them to juelich_tests and label them with "DEV", ja?

**Comment by chraibi** [comment]
> Please add the required python dependencies to `container/build/Dockerfile`

done!

**Comment by schroedtert** [comment]
> > Why do you call the tests develop_test? What do you mean with develop?
> 
> These are featurs still WIP.
> (being **develop**ed)
> 
> But probably we can move them to juelich_tests and label them with "DEV", ja?

What features are still WIP? The tests?

**Comment by chraibi** [comment]
> What features are still WIP?

In my opinion at least these:

- Sources (now more or less I'm more confident in the algorithm 😃 )
- Trains (definitely buggy)
- Waiting area 🙄
- Door FLow Regulation
- Schedules

**Comment by schroedtert** [comment]
> > What features are still WIP?
> 
> In my opinion at least these:
> 
>     * Sources (now more or less I'm more confident in the algorithm smiley )
> 
>     * Trains (definitely buggy)
> 
>     * Waiting area roll_eyes
> 
>     * Door FLow Regulation
> 
>     * Schedules

But in this PR you only work on sources, why no confidence in including the tests to our normal tests?

File: systemtest/JPSRunTest.py [hunk 1]
```diff
@@ -220,10 +220,17 @@ def __execute_test(self, executable, inifile, testfunction, *args):
...
         logging.info('end simulation ...\n--------------\n')
-        trajfile = os.path.join("trajectories", "traj" + inifile.split("ini")[2])
-        logging.info('trajfile = <%s>', trajfile)
+        filename, file_extension = os.path.splitext(inifile)
+        logging.info("inifile <%s>", filename)
+        trajfile = os.path.join("trajectories", "traj" + file_extension)
         if not path.exists(trajfile):
-            logging.critical("trajfile <%s> does not exist", trajfile)
-            exit(self.FAILURE)
+            trajfile = os.path.join("trajectories", "traj.txt")
```
Comment (chraibi) [inline systemtest/JPSRunTest.py]: Oh wow! BOOOMER!!!!! 😱

File: container/build/Dockerfile [hunk 2]
```diff
@@ -28,4 +28,6 @@ RUN cd /opt && CXX=/usr/bin/g++-8 CC=/usr/bin/gcc-8 ./setup-deps.sh
 RUN pip3 install numpy && \
     pip3 install matplotlib && \
     pip3 install pandas && \
-    pip install numpy
\ No newline at end of file
+    pip install numpy && \
+    pip3 install numpy && \
+    pip3 install dataclasses
```
Comment (schroedtert) [inline container/build/Dockerfile]: why is `numpy` installed twice with `pip3`?

File: docs/pages/jpscore/jpscore_introduction.md [hunk 3]
```diff
@@ -4,12 +4,12 @@ keywords: simulation
 tags: [jpscore, getting_started]
 sidebar: jupedsim_sidebar
 permalink: jpscore_introduction.html
-summary: jpscore is the simulation module of JuPedSim. It's a command line tool to simulate the evacuate of pedestrians in continous space.
+summary: jpscore is the simulation module of JuPedSim. It's a command line tool to simulate the evacuate of pedestrians in continuous space.
```
Comment (schroedtert) [inline docs/pages/jpscore/jpscore_introduction.md]: Why are these typo corrections included in a PR related to sources? Please move to a seperate PR

File: docs/pages/jpscore/jpscore_schedule.md [hunk 4]
```diff
@@ -11,7 +11,7 @@ last_updated: Dec 31, 2019
 
 ## Definition
 
-In general a definition of a valide schedule is composed of two parts:
+In general a definition of a valid schedule is composed of two parts:
```
Comment (schroedtert) [inline docs/pages/jpscore/jpscore_schedule.md]: Does not belong here (see above)

File: docs/pages/jpscore/jpscore_trains.md [hunk 5]
```diff
@@ -5,14 +5,14 @@ tags: [jpscore, file, simulation]
 sidebar: jupedsim_sidebar
 folder: jpscore
 permalink: jpscore_trains.html
-summary: The interaction with trains is modeled with an event-based deleting and creating of doors. When a train arrives on the plattform, doors are created and again deleting when the train departs.
+summary: The interaction with trains is modeled with an event-based deleting and creating of doors. When a train arrives on the platform, doors are created and again deleting when the train departs.
```
Comment (schroedtert) [inline docs/pages/jpscore/jpscore_trains.md]: Does not belong here (see above)

File: libcore/src/pedestrian/AgentsSourcesManager.cpp [hunk 6]
```diff
@@ -101,11 +99,16 @@ bool AgentsSourcesManager::ProcessAllSources() const
         // inTime is always true if src got some PlanTime (default values
         // if src has no PlanTime, then this is set to 0. In this case inTime
         // is important in the following condition
```
Comment (schroedtert) [inline libcore/src/pedestrian/AgentsSourcesManager.cpp]: What is `PlanTime`?

File: systemtest/develop_tests/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,4 @@
+add_test(NAME develop_doors_test-1 COMMAND ${PYTHON_EXECUTABLE} ${CMAKE_SOURCE_DIR}/systemtest/develop_tests/doors/runtest_dev_1.py)
+set_tests_properties(
+    develop_doors_test-1
+    PROPERTIES LABELS "CI:MEDIUM")
```
Comment (schroedtert) [inline systemtest/develop_tests/CMakeLists.txt]: Why included in this PR? Does not seem to be related to sources.
Is the test `doors/runtest_dev_1.py` included in this PR?

File: systemtest/develop_tests/README.md [hunk 8]
```diff
@@ -0,0 +1,7 @@
+# About
+
+These tests are about features, that are being developed and need thourough testing.
+
+# Doors
+
+Testing if flow regulation at doors is correct by testing agains "door statistics".
```
Comment (schroedtert) [inline systemtest/develop_tests/README.md]: Why included in this PR? Does not seem to be related to sources

File: systemtest/juelich_tests/README.md [hunk 9]
```diff
@@ -1,11 +1,11 @@
 # Juelich Tests
 Note that a detailed description of the test cases can be found at
-<http://www.jupedsim.org/jpscore/2016-11-02-juelich>
+http://www.jupedsim.org/jpscore_tests_juelich.html
```
Comment (schroedtert) [inline systemtest/juelich_tests/README.md]: Does to seem to be related to sources (see above)

File: systemtest/source_tests/test_source_1/runtest_source_1.py [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+    logging.info(debug_msg)
+    times = []
+    # TODO
+    # if s.N_create:
+    #     correct_length = s.N_create == len(s.ids)
+    #     if no correct_length:
+    #         logging.error("expected "+ str(s.N_create)
+    #                       ", but got "+str(len(s.ids))
+    #                       )
+    #         return False
```
Comment (schroedtert) [inline systemtest/source_tests/test_source_1/runtest_source_1.py]: Is this important?

File: systemtest/source_tests/test_source_1/runtest_source_1.py [hunk 11]
```diff
@@ -0,0 +1,297 @@
+#!/usr/bin/env python3
```
Comment (schroedtert) [inline systemtest/source_tests/test_source_1/runtest_source_1.py]: What do you test here? Could you write a readme at the top of the file? That would it make easier to review.

File: docs/pages/jpscore/jpscore_sources.md [hunk 12]
```diff
@@ -11,30 +11,74 @@ last_updated: Dec 31, 2019
...
+
+- `rate` (int): rate of generation of `percent`$$\times$$`N_create` agents. (default: `frequency`)
+- `time_min` (float), `time_max`: Time lifespan of the source.
+- `agents_max` (int): maximal number of agents produced by that source. (default: 10)
 - `group_id`: group id of the agents. This `id` **should match** a predefined group in the section [Agents_distribution](jpscore_inifile.html#agents_distribution).
-- `time`: time of appearance of agent with id `agent_id`. Here `agents_max=1`.
-- `startX`, `startY`: Distribute one agent at a fix position.
-- `x_min`, `x_max`, `y_min`, `y_max`: Bounding box for generation of agents.
-- `greedy` (default `false`): returns a Voronoi vertex randomly with respect to weights proportional to squared distances.
+- `time` (float): time of appearance of agent. Should be used together with `agent_id`. If used then `agents_max=1` and `frequency=1`.
```
Comment (schroedtert) [inline docs/pages/jpscore/jpscore_sources.md]: What is this `agent_id`? Is not explained anywhere. If I understand this correctly, setting `agent_id` will just create an agent with the specific id at time=`time`?

File: systemtest/CMakeLists.txt [hunk 13]
```diff
@@ -2,3 +2,5 @@ add_subdirectory(test_geometry)
 add_subdirectory(juelich_tests)
 add_subdirectory(rimea_tests)
 add_subdirectory(router_tests)
+add_subdirectory(develop_tests)
```
Comment (schroedtert) [inline systemtest/CMakeLists.txt]: Does this belong to this PR, as only a test for doors is later added/explained?

File: systemtest/juelich_tests/test_15/master_ini.xml [hunk 14]
```diff
@@ -2,14 +2,14 @@
...
   <max_sim_time unit="sec">2000</max_sim_time>
   <!-- geometry file -->
   <geometry>../geometry.xml</geometry>
   <!-- traectories file and format -->
   <trajectories format="xml-plain" embed_mesh="false" fps="8">
-    <file location="Traj_test_1.xml" />
+    <file location="Traj_test_15.xml" />
   </trajectories>
   <!-- where to store the logs -->
   <logfile>log</logfile>
```
Comment (schroedtert) [inline systemtest/juelich_tests/test_15/master_ini.xml]: Do we still need this test? We should talk about it on monday

File: container/build/Dockerfile (hunk 2 above)
Comment (schroedtert) [reply container/build/Dockerfile]: Still needs fix

File: systemtest/source_tests/test_source_1/runtest_source_1.py (hunk 11 above)
Comment (chraibi) [reply systemtest/source_tests/test_source_1/runtest_source_1.py]: I think this is now solved.

File: docs/pages/jpscore/jpscore_introduction.md (hunk 3 above)
Comment (chraibi) [reply docs/pages/jpscore/jpscore_introduction.md]: removed from this PR
